## KPI in Covid19
   1.  Based on age groups the number of cases, deaths and intensive care patients in Sweden.
   2.  On which date what region has registered highest number of cases.
   3.  Number of deaths occured in each region of Sweden.

## # KPI 1:

  - Read the data from corresponding sheet
  - From different age groups data was collected
  - Data is represented graphically

In [418]:
import pandas as pd
import plotly_express as px
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_excel("../Labb/Data/Covid19.xlsx",sheet_name="Totalt antal per åldersgrupp")  
df.head()

Åldersgrupp  Totalt_antal_fall  Totalt_antal_intensivvårdade  \
0    Ålder_0_9             138071                           109   
1  Ålder_10_19             355823                           101   
2  Ålder_20_29             418506                           285   
3  Ålder_30_39             493443                           492   
4  Ålder_40_49             474702                           997   

   Totalt_antal_avlidna  
0                    17  
1                     9  
2                    41  
3                    71  
4                   172

In [419]:
fig = px.line(
    df,
    x="Åldersgrupp",
    y=["Totalt_antal_fall", "Totalt_antal_intensivvårdade", "Totalt_antal_avlidna"],
    labels = {"Åldersgrupp":"Age groups","value":"Population", "variable": "Covid19"},
    log_y=True,
    title="Number of cases, intensive care patients and deaths in Sweden",
)
fig.show()
#fig.write_html("uppgift 3-1.html", auto_open = True)

##  # KPI 2 : 

   - Read the data related to  corresponding sheet on Covid19 
   - Pick out the date by find out maximum number of cases from all region
   - With that maximum cases we can pick out the name of region


In [420]:
df = pd.read_excel("../Labb/Data/Covid19.xlsx",sheet_name="Antal per dag region")  
df.head()

Statistikdatum  Totalt_antal_fall  Blekinge  Dalarna  Gotland  Gävleborg  \
0     2020-02-04                  1         0        0        0          0   
1     2020-02-05                  0         0        0        0          0   
2     2020-02-06                  0         0        0        0          0   
3     2020-02-07                  0         0        0        0          0   
4     2020-02-08                  0         0        0        0          0   

   Halland  Jämtland_Härjedalen  Jönköping  Kalmar  ...  Stockholm  Sörmland  \
0        0                    0          1       0  ...          0         0   
1        0                    0          0       0  ...          0         0   
2        0                    0          0       0  ...          0         0   
3        0                    0          0       0  ...          0         0   
4        0                    0          0       0  ...          0         0   

   Uppsala  Värmland  Västerbotten  Västernorrland  Västmanland  \
0        0         0             0               0            0   
1        0         0             0               0            0   
2        0         0             0               0            0   
3        0         0             0               0            0   
4        0         0             0               0            0   

   Västra_Götaland  Örebro  Östergötland  
0                0       0             0  
1                0       0             0  
2                0       0             0  
3                0       0             0  
4                0       0             0  

[5 rows x 23 columns]

In [421]:
#df.query('Totalt_antal_fall == Totalt_antal_fall.max()') # It is also working

df.nlargest(2,['Totalt_antal_fall']) # to find how many top rows you want fo specific column

Statistikdatum  Totalt_antal_fall  Blekinge  Dalarna  Gotland  Gävleborg  \
723     2022-01-27              53881       547     1371      235       1496   
721     2022-01-25              46669       445     1580      336       1284   

     Halland  Jämtland_Härjedalen  Jönköping  Kalmar  ...  Stockholm  \
723     2878                 1232       2682    1259  ...       8169   
721     2056                  492       1461    1562  ...       9642   

     Sörmland  Uppsala  Värmland  Västerbotten  Västernorrland  Västmanland  \
723       745     2454      1474           753            1178         2151   
721      1579     2033      2147          1230             922         2093   

     Västra_Götaland  Örebro  Östergötland  
723            14429    1464          2139  
721             8609    2245          1564  

[2 rows x 23 columns]

In [422]:
# returns row with maximum value in "Totalt_antal_fall" column
df_max = df[df["Totalt_antal_fall"] == df["Totalt_antal_fall"].max()]
df_max


Statistikdatum  Totalt_antal_fall  Blekinge  Dalarna  Gotland  Gävleborg  \
723     2022-01-27              53881       547     1371      235       1496   

     Halland  Jämtland_Härjedalen  Jönköping  Kalmar  ...  Stockholm  \
723     2878                 1232       2682    1259  ...       8169   

     Sörmland  Uppsala  Värmland  Västerbotten  Västernorrland  Västmanland  \
723       745     2454      1474           753            1178         2151   

     Västra_Götaland  Örebro  Östergötland  
723            14429    1464          2139  

[1 rows x 23 columns]

In [423]:
df_region = df_max.loc[:,  ~df_max.columns.isin(['Statistikdatum','Totalt_antal_fall'])] # Exlude multiple columns
df_region


Blekinge  Dalarna  Gotland  Gävleborg  Halland  Jämtland_Härjedalen  \
723       547     1371      235       1496     2878                 1232   

     Jönköping  Kalmar  Kronoberg  Norrbotten  ...  Stockholm  Sörmland  \
723       2682    1259       1239         470  ...       8169       745   

     Uppsala  Värmland  Västerbotten  Västernorrland  Västmanland  \
723     2454      1474           753            1178         2151   

     Västra_Götaland  Örebro  Östergötland  
723            14429    1464          2139  

[1 rows x 21 columns]

In [424]:
region = df_region.idxmax(axis = 1)
print(region)

723    Västra_Götaland
dtype: object


###  --> The heighest number of cases  are registerd in Västra_Götaland region on 2022-01-27	

## # KPI 3:

  - Read data 
  - collect the data of number of deaths caused by Covid19 in each region of Sweden
  - Represents graph deaths per region
  - plot graphically

In [425]:
df = pd.read_excel("../Labb/Data/Covid19.xlsx",sheet_name="Totalt antal per region")  
df.head()

Region  Totalt_antal_fall  Fall_per_100000_inv  \
0   Blekinge              30829         19371.449951   
1    Dalarna              75091         26098.780273   
2    Gotland              11874         19776.671875   
3  Gävleborg              74803         26020.503418   
4    Halland             108822         32349.047119   

   Totalt_antal_intensivvårdade  Totalt_antal_avlidna  
0                            85                   184  
1                           260                   544  
2                            46                    82  
3                           332                   754  
4                           229                   518

In [426]:
fig = px.line(
    data_frame=df,
    
    x="Region",
   
    y="Totalt_antal_avlidna",
    
    labels={"Region": "Region", "Totalt_antal_avlidna": "Number of deaths"},
    title="Number of Deaths occured by Covid19 in each region of Sweden",
    
    )
fig.show()
#fig.write_html("uppgift 3-3.html", auto_open = True)


##  KPI in Vaccine

 4.  Number of doses taken by different age groups.
 5.  At what week and how many persons got 4 doses. 
 6.  All 5doses taken in each community.



## # KPI 4:

- Read data from Vaccine with different sheets
- Collect the data of all doses per age group from those sheets
- Merge the collected data
- Plot the graph


In [427]:
df_vaccine = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade kommun och ålder")
df_vaccine.head()

Län        Län_namn  Kommun     Kommun_namn  Ålder  Befolkning  \
0    1  Stockholms län     114  Upplands Väsby  12-15        2422   
1    1  Stockholms län     114  Upplands Väsby  16-17        1203   
2    1  Stockholms län     114  Upplands Väsby  18-29        6692   
3    1  Stockholms län     114  Upplands Väsby  30-39        7332   
4    1  Stockholms län     114  Upplands Väsby  40-49        6946   

   Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser  Antal 4 doser  \
0               1206                 1046            NaN            NaN   
1                839                  755            NaN            NaN   
2               4887                 4469         1959.0            NaN   
3               5542                 5240         2878.0            NaN   
4               5592                 5429         3719.0            NaN   

   Andel minst 1 dos  Andel minst 2 doser  Andel 3 doser  Andel 4 doser  
0           0.497936             0.431874            NaN            NaN  
1           0.697423             0.627598            NaN            NaN  
2           0.730275             0.667812       0.292738            NaN  
3           0.755865             0.714675       0.392526            NaN  
4           0.805068             0.781601       0.535416            NaN

In [428]:
# For doses 1 to 3
df = df_vaccine.groupby("Ålder")[["Antal minst 1 dos","Antal minst 2 doser", "Antal 3 doser"]].sum(numeric_only= True).reset_index()
df1_to_3= df.rename(columns={"Ålder":"Åldersgrupp"}) # renamed bcz to maintain same name 
df1_to_3.replace("60-69","65-69", inplace=True) # other data having age 65-69 
df1_to_3


Åldersgrupp  Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser
0           12-15             278405               255541            0.0
1           16-17             187077               175480            0.0
2           18-29            1201569              1144897       573532.0
3           30-39            1190083              1148376       689647.0
4           40-49            1117660              1095826       791299.0
5           50-59            1210791              1197415      1000554.0
6           65-69            1044061              1036035       949409.0
7           70-79             990440               985748       951943.0
8           80-89             480671               479037       468149.0
9  90 eller äldre             109619               109230       106717.0

In [429]:
# for 4th dose
df4 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Dos 4 per åldersgrupp")
df4.head()

Region     Åldersgrupp  Antal vaccinerade  Andel vaccinerade  \
0  | Sverige |           65-69             376002           0.689459   
1  | Sverige |           70-79             828445           0.801892   
2  | Sverige |           80-89             428931           0.863475   
3  | Sverige |  90 eller äldre              97491           0.854128   
4  | Sverige |          Totalt            1730869           0.790581   

  Vaccinationsstatus  
0            4 doser  
1            4 doser  
2            4 doser  
3            4 doser  
4            4 doser

In [430]:
df4 = df4.groupby("Åldersgrupp")[["Antal vaccinerade"]].max().reset_index()
df4 = df4.rename(columns={"Antal vaccinerade":"Antal 4 doser"})
df4

Åldersgrupp  Antal 4 doser
0           65-69         376002
1           70-79         828445
2           80-89         428931
3  90 eller äldre          97491
4          Totalt        1730869

In [431]:
# for 5th dose
df5 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Dos 5 per åldersgrupp")
df5.head()

Region     Åldersgrupp  Antal vaccinerade  Andel vaccinerade  \
0  | Sverige |           65-69             194116           0.355942   
1  | Sverige |           70-79             513081           0.496636   
2  | Sverige |           80-89             309976           0.624008   
3  | Sverige |  90 eller äldre              66927           0.586354   
4  | Sverige |          Totalt            1084100           0.495167   

  Vaccinationsstatus  
0            5 doser  
1            5 doser  
2            5 doser  
3            5 doser  
4            5 doser

In [432]:
df5= df5.groupby("Åldersgrupp")[["Antal vaccinerade"]].max().reset_index()
df5 = df5.rename(columns={"Antal vaccinerade":"Antal 5 doser"}) # change name as above doses
df5

Åldersgrupp  Antal 5 doser
0           65-69         194116
1           70-79         513081
2           80-89         309976
3  90 eller äldre          66927
4          Totalt        1084100

In [433]:
# Merge all the doses
df_total = df1_to_3.merge(df4,on="Åldersgrupp",how = "left").merge(df5,on="Åldersgrupp",how = "left")
df_total

Åldersgrupp  Antal minst 1 dos  Antal minst 2 doser  Antal 3 doser  \
0           12-15             278405               255541            0.0   
1           16-17             187077               175480            0.0   
2           18-29            1201569              1144897       573532.0   
3           30-39            1190083              1148376       689647.0   
4           40-49            1117660              1095826       791299.0   
5           50-59            1210791              1197415      1000554.0   
6           65-69            1044061              1036035       949409.0   
7           70-79             990440               985748       951943.0   
8           80-89             480671               479037       468149.0   
9  90 eller äldre             109619               109230       106717.0   

   Antal 4 doser  Antal 5 doser  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4            NaN            NaN  
5            NaN            NaN  
6       376002.0       194116.0  
7       828445.0       513081.0  
8       428931.0       309976.0  
9        97491.0        66927.0

In [434]:
fig = px.bar(
    data_frame=df_total,
    
    x="Åldersgrupp",
    
    y=["Antal minst 1 dos", "Antal minst 2 doser", "Antal 3 doser", "Antal 4 doser", "Antal 5 doser"],
    
    labels={"Åldersgrupp": "Age groups", "value": "Population", "variable":"Doses"},
    title="Number of doses taken by different age groups",
    barmode="group"
    

)
fig.show()
#fig.write_html("uppgift 3-4.html", auto_open = True)

## #KPI 5:

- Read data from Vaccine 
- Check the maximum value of vaccinated in Sweden
- From that we get week and total number of persons


In [435]:
df = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade tidsserie dos 4")
df.head()

Vecka    År        Region  Antal vaccinerade  Andel vaccinerade  \
0      3  2022   | Sverige |                 20           0.000002   
1      3  2022     Stockholm                  0           0.000000   
2      3  2022       Uppsala                  0           0.000000   
3      3  2022  Södermanland                  5           0.000021   
4      3  2022  Östergötland                  0           0.000000   

  Vaccinationsstatus  
0            4 doser  
1            4 doser  
2            4 doser  
3            4 doser  
4            4 doser

In [436]:
df[df["Antal vaccinerade"] == df["Antal vaccinerade"].max()]

Vecka    År       Region  Antal vaccinerade  Andel vaccinerade  \
858     42  2022  | Sverige |            2427338           0.290789   

    Vaccinationsstatus  
858            4 doser

### --> The total number of persons in Sweden took 4 doses based on dataset  are 2427338 on Vecka42 2022

## # KPI 6:

- Read data from Vaccine with diffenent number of sheets
- Collect data with commun have all doses from different sheets
- Merge data of all communities with 5 doses
- Plot graph

In [437]:
# data for all communities have 1 and 2 doser
df1 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade kommun") 
df1.head()

KnKod          KnNamn  Antal_dos1  Antal_dos2  Andel_dos1  Andel_dos2
0    114  Upplands Väsby       33514       32224    0.808677    0.777550
1    115      Vallentuna       25425       24744    0.864443    0.841289
2    117       Österåker       36270       35217    0.870766    0.845485
3    120          Värmdö       34602       33528    0.870688    0.843663
4    123        Järfälla       57901       55476    0.806073    0.772313

In [438]:
# data for all communities have  3rd dose
df2 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade kommun dos 3")
df2.head()

KnKod          KnNamn  Antal_dos3  Andel_dos3
0    114  Upplands Väsby       22131    0.585198
1    115      Vallentuna       18072    0.687436
2    117       Österåker       26146    0.694966
3    120          Värmdö       24513    0.683823
4    123        Järfälla       36822    0.562726

In [439]:
#data for all communities have  4th dose
df3 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade kommun dos 4")
df3.head()

KnKod          KnNamn  Antal_dos4  Andel_dos4
0    114  Upplands Väsby        9719    0.256994
1    115      Vallentuna        8144    0.309787
2    117       Österåker       11995    0.318829
3    120          Värmdö       10785    0.300862
4    123        Järfälla       15401    0.235363

In [440]:
#data for all communities have 5th dose
df4 = pd.read_excel("../Labb/Data/Vaccine.xlsx", sheet_name="Vaccinerade kommun dos 5")
df4.head()

KnKod          KnNamn  Antal_dos5  Andel_dos5
0    114  Upplands Väsby        3714    0.452982
1    115      Vallentuna        2877    0.503588
2    117       Österåker        4635    0.506613
3    120          Värmdö        4317    0.505030
4    123        Järfälla        5876    0.422126

In [441]:
# Merge all the data into one
df_merge = df1.merge(df2,on=["KnKod","KnNamn"]).merge(df3,on=["KnKod","KnNamn"]).merge(df4, on=["KnKod","KnNamn"])
df_merge.head()

KnKod          KnNamn  Antal_dos1  Antal_dos2  Andel_dos1  Andel_dos2  \
0    114  Upplands Väsby       33514       32224    0.808677    0.777550   
1    115      Vallentuna       25425       24744    0.864443    0.841289   
2    117       Österåker       36270       35217    0.870766    0.845485   
3    120          Värmdö       34602       33528    0.870688    0.843663   
4    123        Järfälla       57901       55476    0.806073    0.772313   

   Antal_dos3  Andel_dos3  Antal_dos4  Andel_dos4  Antal_dos5  Andel_dos5  
0       22131    0.585198        9719    0.256994        3714    0.452982  
1       18072    0.687436        8144    0.309787        2877    0.503588  
2       26146    0.694966       11995    0.318829        4635    0.506613  
3       24513    0.683823       10785    0.300862        4317    0.505030  
4       36822    0.562726       15401    0.235363        5876    0.422126

In [442]:
fig = px.line(
    data_frame=df_merge,
    
    x="KnNamn",
    
    y=["Antal_dos1", "Antal_dos2", "Antal_dos3", "Antal_dos4","Antal_dos5"],
    
    labels={"KnNamn": "Commun name", "value": "Population", "variable":"Doses"},
    title="Total number of doses taken in each community",
    log_y= True,
    
)
fig.show()
#fig.write_html("uppgift 3-6.html", auto_open = True)